File: research/python/analysis/audio_inspector_gui.py
import tkinter as tk
from tkinter import filedialog, messagebox
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.backends.backend_tkagg import FigureCanvasTkAgg, NavigationToolbar2Tk
from scipy.io import wavfile
from scipy.signal import welch
import os

class AudioInspectorApp:
    def __init__(self, root):
        self.root = root
        self.root.title("Audio Inspector - Interactive PSD & Envelope")
        self.root.geometry("1000x800")
        
        # --- Top Frame for Controls ---
        control_frame = tk.Frame(self.root, pady=10, padx=10)
        control_frame.pack(side=tk.TOP, fill=tk.X)
        
        self.load_btn = tk.Button(control_frame, text="Load WAV File", command=self.load_file, font=("Arial", 12, "bold"))
        self.load_btn.pack(side=tk.LEFT)
        
        self.file_label = tk.Label(control_frame, text="No file loaded", font=("Arial", 10), fg="gray")
        self.file_label.pack(side=tk.LEFT, padx=15)
        
        # --- Matplotlib Figure Setup ---
        self.fig, (self.ax_env, self.ax_psd) = plt.subplots(2, 1, figsize=(10, 8))
        self.fig.tight_layout(pad=4.0)
        
        self.canvas = FigureCanvasTkAgg(self.fig, master=self.root)
        self.canvas.draw()
        
        # Add navigation toolbar (zoom, pan, save)
        toolbar_frame = tk.Frame(self.root)
        toolbar_frame.pack(side=tk.TOP, fill=tk.X)
        self.toolbar = NavigationToolbar2Tk(self.canvas, toolbar_frame)
        self.toolbar.update()
        
        self.canvas.get_tk_widget().pack(side=tk.TOP, fill=tk.BOTH, expand=1)
        
    def load_file(self):
        filepath = filedialog.askopenfilename(
            title="Select WAV File",
            filetypes=(("WAV files", "*.wav"), ("All files", "*.*"))
        )
        if not filepath:
            return
            
        self.file_label.config(text=f"Loading: {os.path.basename(filepath)}...", fg="blue")
        self.root.update()
        
        try:
            self.process_and_plot(filepath)
            self.file_label.config(text=f"Loaded: {os.path.basename(filepath)}", fg="green")
        except Exception as e:
            messagebox.showerror("Error", f"Failed to load audio file:\n{str(e)}")
            self.file_label.config(text="Error loading file", fg="red")
            
    def process_and_plot(self, filepath):
        # 1. Load Audio
        sr, data = wavfile.read(filepath)
        
        # Convert to mono if stereo
        if len(data.shape) > 1:
            data = data.mean(axis=1)
            
        # 2. Calculate Fast Envelope (Downsampled Rolling Max)
        # To keep UI responsive, we calculate envelope by taking max over chunks
        chunk_size = sr // 100 # 100 envelope points per second
        num_chunks = len(data) // chunk_size
        
        if num_chunks > 0:
            truncated_data = data[:num_chunks * chunk_size]
            reshaped_data = truncated_data.reshape((num_chunks, chunk_size))
            envelope = np.max(np.abs(reshaped_data), axis=1)
            time_env = np.arange(num_chunks) * (chunk_size / sr)
        else:
            envelope = np.abs(data)
            time_env = np.arange(len(data)) / sr

        # 3. Calculate Welch PSD
        nperseg = min(4096, len(data))
        freqs, psd = welch(data, fs=sr, nperseg=nperseg)
        psd_db = 10 * np.log10(psd + 1e-12)
        
        # 4. Update Plots
        self.ax_env.clear()
        self.ax_env.plot(time_env, envelope, color='dodgerblue')
        self.ax_env.set_title("Time-Domain Envelope (Downsampled Absolute Max)", fontweight='bold')
        self.ax_env.set_xlabel("Time (s)")
        self.ax_env.set_ylabel("Amplitude")
        self.ax_env.grid(True, alpha=0.3)
        
        self.ax_psd.clear()
        self.ax_psd.semilogx(freqs, psd_db, color='crimson')
        self.ax_psd.set_title("Welch Power Spectral Density (PSD)", fontweight='bold')
        self.ax_psd.set_xlabel("Frequency (Hz)")
        self.ax_psd.set_ylabel("Power (dB/Hz)")
        self.ax_psd.grid(True, which='both', alpha=0.3)
        
        self.fig.tight_layout(pad=4.0)
        self.canvas.draw()

if __name__ == "__main__":
    root = tk.Tk()
    app = AudioInspectorApp(root)
    root.mainloop()

File: research/matlab/audio_inspector_gui.m
function audio_inspector_gui()
    % Create UI Figure
    fig = uifigure('Name', 'Audio Inspector', 'Position', [100 100 800 800]);
    
    % Add Load Button
    btn = uibutton(fig, 'Position', [20, 750, 100, 30], 'Text', 'Load WAV File', ...
        'ButtonPushedFcn', @(btn,event) loadAudioFile(fig));
    
    % Status Label
    statusLbl = uilabel(fig, 'Position', [130, 750, 600, 30], 'Text', 'No file loaded.');
    
    % Add Axes for Envelope
    axEnv = uiaxes(fig, 'Position', [50, 500, 650, 200]);
    title(axEnv, 'Time-Domain Envelope (Downsampled Absolute Max)');
    xlabel(axEnv, 'Time (s)');
    ylabel(axEnv, 'Amplitude');
    grid(axEnv, 'on');
    
    % Add Axes for Spectrogram
    axSpec = uiaxes(fig, 'Position', [50, 275, 650, 200]);
    title(axSpec, 'Spectrogram');
    xlabel(axSpec, 'Time (s)');
    ylabel(axSpec, 'Frequency (Hz)');
    
    % Add Axes for PSD
    axPSD = uiaxes(fig, 'Position', [50, 50, 650, 200]);
    title(axPSD, 'Welch Power Spectral Density (PSD)');
    xlabel(axPSD, 'Frequency (Hz)');
    ylabel(axPSD, 'Power/Frequency (dB/Hz)');
    grid(axPSD, 'on');
    
    % Store components in UserData for callback access
    fig.UserData = struct('axEnv', axEnv, 'axSpec', axSpec, 'axPSD', axPSD, 'statusLbl', statusLbl);
end

function loadAudioFile(fig)
    [file, path] = uigetfile('*.wav', 'Select WAV File');
    if isequal(file, 0)
        return;
    end
    
    filepath = fullfile(path, file);
    fig.UserData.statusLbl.Text = ['Loading: ', file, '...'];
    drawnow;
    
    try
        [data, fs] = audioread(filepath);
        
        % Convert to mono if stereo
        if size(data, 2) > 1
            data = mean(data, 2);
        end
        
        % Calculate Fast Envelope (Downsampled Rolling Max)
        chunk_size = floor(fs / 100); % 100 points per second
        num_chunks = floor(length(data) / chunk_size);
        
        if num_chunks > 0
            truncated_data = data(1:(num_chunks * chunk_size));
            reshaped_data = reshape(truncated_data, chunk_size, num_chunks);
            env = max(abs(reshaped_data), [], 1);
            time_env = (0:(num_chunks-1)) * (chunk_size / fs);
        else
            env = abs(data);
            time_env = (0:(length(data)-1)) / fs;
        end
        
        % Plot Envelope
        plot(fig.UserData.axEnv, time_env, env, 'b');
        title(fig.UserData.axEnv, 'Time-Domain Envelope (Downsampled Absolute Max)');
        
        % Calculate and Plot Spectrogram
        nfft = min(8192, length(data)); % Increased for higher frequency resolution
        noverlap = round(nfft * 0.875); % 87.5% overlap for higher time resolution
        [~, f_spec, t_spec, p] = spectrogram(data, nfft, noverlap, nfft, fs);
        p_db = 10 * log10(abs(p) + 1e-12);
        imagesc(fig.UserData.axSpec, t_spec, f_spec, p_db);
        axis(fig.UserData.axSpec, 'xy');
        title(fig.UserData.axSpec, 'Spectrogram');
        xlabel(fig.UserData.axSpec, 'Time (s)');
        ylabel(fig.UserData.axSpec, 'Frequency (Hz)');
        colorbar(fig.UserData.axSpec);
        
        % Calculate Welch PSD
        % Ensure data is a vector
        nperseg = min(4096, length(data));
        [pxx, f] = pwelch(data, nperseg, [], [], fs);
        pxx_db = 10 * log10(pxx + 1e-12);
        
        % Plot PSD
        semilogx(fig.UserData.axPSD, f, pxx_db, 'r');
        title(fig.UserData.axPSD, 'Welch Power Spectral Density (PSD)');
        
        fig.UserData.statusLbl.Text = ['Loaded: ', file];
    catch e
        uialert(fig, ['Error loading file: ', e.message], 'Error');
        fig.UserData.statusLbl.Text = 'Error loading file.';
    end
end
